File: parser/src/main/java/org/cooklang/parser/Recipe.java
package org.cooklang.parser;

import java.util.ArrayList;
import java.util.Hashtable;
import java.util.List;
import java.util.Map;

public class Recipe {
    public List<Step> steps = new ArrayList<>();
    public Hashtable<String, String> metadata = new Hashtable<String, String>();

    public void addStep(Step step) {
        steps.add(step);
    }

    public void addMetadata(String key, String value) {
        metadata.put(key, value);
    }

    public String toString() {
        if (steps == null || steps.size() <= 0) return "";

        StringBuilder sb = new StringBuilder();

        for (Map.Entry<String, String> entry : metadata.entrySet()) {
            String key = entry.getKey();
            String value = entry.getValue();

            sb.append(String.format(">> %s: %s\n", key, value));
        }

        for (int i = 0; i < steps.size(); i++) {

            sb.append(steps.get(i).toString());

            // if not the last item
            if (i != steps.size() - 1) {
                sb.append("\n");
            }

        }

        return sb.toString();

    }
}

File: parser/src/main/java/org/cooklang/parser/Recipe.kt
package org.cooklang.parser

import java.lang.StringBuilder
import java.util.*

class Recipe : Displayable {
    val steps: MutableList<Step> = mutableListOf()
    var metadata = Hashtable<String, String>()
    fun addStep(step: Step) {
        steps.add(step)
    }

    val ingredients: List<Ingredient> get() = steps.map { it.ingredients }.reduce { acc, list -> acc + list }
    val cookware: List<Cookware> get() = steps.map { it.cookware }.reduce { acc, list -> acc + list }

    fun addMetadata(key: String, value: String) {
        metadata[key] = value
    }

    override fun getDisplayString(): String {
        if (steps.size <= 0) return ""
        val sb = StringBuilder()
        for ((key, value) in metadata) {
            sb.append(">> $key: $value\n")
        }
        steps.forEachIndexed { index, step ->
            sb.append("${index + 1}. ${step.getDisplayString()}")
            if (index != steps.lastIndex) {
                sb.append("\n")
            }
        }
        return sb.toString()
    }
}